File: 23-24/16/nov-15.py
def permute(str, left, right, arr):
    if left == right:
        arr.append(str)
    else:
        for i in range(left, right):
            str[i] = str[right]
            permute(str, left, right, arr)
            str[right] = str[i]
        

str = "ABC"
results = []
permute(str, 0, len(str)-1, results)
print(results)

File: 23-24/16/nov-15.ts
function permute(chars: string[], left: number, right: number, arr: string[]) {
    if (left === right) {
        arr.push(chars.join(""));
        return;
    }

    for (let i = left; i <= right; i++) {
        const tmp = chars[i];
        chars[left] = chars[i];
        chars[i] = tmp;
        permute(chars, left + 1, right, arr);
        // TODO: swap back
    }
}

const str = "XY";
const results = [];
permute(str.split(""), 0, str.length - 1, results);
console.log(results);
